File: sdks/ios/ErrorTracker/ErrorTracker.swift
import Foundation
import UIKit

/**
 * Error Tracker SDK for iOS
 * 
 * Usage:
 * ```
 * ErrorTracker.shared.initialize(apiKey: "err_abc123...", apiUrl: "https://api.example.com/api")
 * ```
 */
public class ErrorTracker {
    public static let shared = ErrorTracker()
    
    private var config: Config?
    private var defaultExceptionHandler: NSUncaughtExceptionHandler?
    
    private init() {}
    
    /**
     * Initialize the SDK
     */
    public func initialize(
        apiKey: String,
        apiUrl: String = "http://localhost:3000/api",
        environment: String? = nil,
        release: String? = nil,
        enabled: Bool = true,
        user: [String: Any]? = nil,
        tags: [String: String]? = nil,
        metadata: [String: Any]? = nil
    ) {
        if config != nil {
            print("[ErrorTracker] SDK already initialized")
            return
        }
        
        if apiKey.isEmpty {
            fatalError("ErrorTracker: apiKey is required")
        }
        
        config = Config(
            apiKey: apiKey,
            apiUrl: apiUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")),
            environment: environment,
            release: release ?? getAppVersion(),
            enabled: enabled,
            user: user ?? [:],
            tags: tags ?? [:],
            metadata: metadata ?? [:]
        )
        
        if enabled {
            setupAutomaticCapture()
        }
    }
    
    /**
     * Capture an error manually
     */
    public func captureError(
        _ error: Error,
        severity: ErrorSeverity = .error,
        metadata: [String: Any]? = nil
    ) {
        guard let config = self.config else {
            print("[ErrorTracker] SDK not initialized. Call initialize() first.")
            return
        }
        
        if !config.enabled {
            return
        }
        
        let errorData = buildErrorData(error: error, severity: severity, metadata: metadata, config: config)
        sendError(errorData, config: config)
    }
    
    /**
     * Capture a message (non-error)
     */
    public func captureMessage(
        _ message: String,
        severity: ErrorSeverity = .info,
        metadata: [String: Any]? = nil
    ) {
        guard let config = self.config else {
            print("[ErrorTracker] SDK not initialized. Call initialize() first.")
            return
        }
        
        if !config.enabled {
            return
        }
        
        let stackTrace = Thread.callStackSymbols.joined(separator: "\n")
        let errorData = ErrorData(
            errorType: "Message",
            message: message,
            stackTrace: stackTrace,
            severity: severity,
            metadata: buildMetadata(config: config, additionalMetadata: metadata)
        )
        
        sendError(errorData, config: config)
    }
    
    /**
     * Set user context
     */
    public func setUser(_ user: [String: Any]) {
        config?.user = user
    }
    
    /**
     * Set tags
     */
    public func setTags(_ tags: [String: String]) {
        guard var config = self.config else {
            print("[ErrorTracker] SDK not initialized. Call initialize() first.")
            return
        }
        config.tags.merge(tags) { (_, new) in new }
        self.config = config
    }
    
    /**
     * Set metadata
     */
    public func setMetadata(_ metadata: [String: Any]) {
        guard var config = self.config else {
            print("[ErrorTracker] SDK not initialized. Call initialize() first.")
            return
        }
        config.metadata.merge(metadata) { (_, new) in new }
        self.config = config
    }
    
    /**
     * Enable/disable the SDK
     */
    public func setEnabled(_ enabled: Bool) {
        config?.enabled = enabled
    }
    
    // MARK: - Private Methods
    
    private func setupAutomaticCapture() {
        // Capture uncaught exceptions
        NSSetUncaughtExceptionHandler { exception in
            ErrorTracker.shared.captureError(
                NSError(domain: exception.name.rawValue, code: 0, userInfo: [
                    NSLocalizedDescriptionKey: exception.reason ?? "Unknown exception"
                ]),
                severity: .critical
            )
        }
        
        // Capture Swift errors (via global error handler if available)
        // Note: Swift errors are harder to catch globally, so manual capture is recommended
    }
    
    private func buildErrorData(
        error: Error,
        severity: ErrorSeverity,
        metadata: [String: Any]?,
        config: Config
    ) -> ErrorData {
        let nsError = error as NSError
        let stackTrace = Thread.callStackSymbols.joined(separator: "\n")
        let errorInfo = extractErrorInfo(error: error, stackTrace: stackTrace)
        
        return ErrorData(
            errorType: errorInfo.errorType,
            message: errorInfo.message,
            stackTrace: stackTrace,
            severity: severity,
            file: errorInfo.file,
            line: errorInfo.line,
            functionName: errorInfo.functionName,
            metadata: buildMetadata(config: config, additionalMetadata: metadata)
        )
    }
    
    private func extractErrorInfo(error: Error, stackTrace: String) -> ErrorInfo {
        let nsError = error as NSError
        let errorType = String(describing: type(of: error))
        let message = nsError.localizedDescription
        
        // Try to extract file, line, and function from stack trace
        let stackLines = stackTrace.components(separatedBy: "\n")
        var file: String? = nil
        var line: Int? = nil
        var functionName: String? = nil
        
        if !stackLines.isEmpty {
            // Pattern: AppName 0x123456 functionName + 123
            let firstLine = stackLines[0]
            if let match = firstLine.range(of: #"(\w+)\s+\+\s+(\d+)"#, options: .regularExpression) {
                functionName = String(firstLine[..<match.lowerBound]).trimmingCharacters(in: .whitespaces)
                if let lineStr = firstLine[match].components(separatedBy: "+").last?.trimmingCharacters(in: .whitespaces),
                   let lineNum = Int(lineStr) {
                    line = lineNum
                }
            }
        }
        
        return ErrorInfo(
            errorType: errorType,
            message: message,
            file: file,
            line: line,
            functionName: functionName
        )
    }
    
    private func buildMetadata(config: Config, additionalMetadata: [String: Any]?) -> [String: Any] {
        var metadata: [String: Any] = [
            "os": "iOS",
            "osVersion": UIDevice.current.systemVersion,
            "deviceModel": UIDevice.current.model,
            "deviceName": UIDevice.current.name,
            "framework": "ios"
        ]
        
        if let environment = config.environment {
            metadata["environment"] = environment
        }
        
        if let release = config.release {
            metadata["release"] = release
        }
        
        if let userId = config.user["id"] {
            metadata["userId"] = userId
        }
        
        if let userName = config.user["username"] {
            metadata["userName"] = userName
        }
        
        if !config.tags.isEmpty {
            metadata["tags"] = config.tags
        }
        
        metadata.merge(config.metadata) { (_, new) in new }
        if let additional = additionalMetadata {
            metadata.merge(additional) { (_, new) in new }
        }
        
        return metadata
    }
    
    private func getAppVersion() -> String? {
        return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
    }
    
    private func sendError(_ errorData: ErrorData, config: Config) {
        let endpoint = "\(config.apiUrl)/errors/report"
        
        guard let url = URL(string: endpoint) else {
            print("[ErrorTracker] Invalid API URL: \(endpoint)")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(config.apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: errorData.toDictionary())
        } catch {
            print("[ErrorTracker] Failed to serialize error data: \(error)")
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                print("[ErrorTracker] Failed to send error: \(error.localizedDescription)")
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
                print("[ErrorTracker] Failed to send error: HTTP \(httpResponse.statusCode)")
            }
        }.resume()
    }
    
    // MARK: - Data Structures
    
    private struct Config {
        let apiKey: String
        let apiUrl: String
        let environment: String?
        let release: String?
        var enabled: Bool
        var user: [String: Any]
        var tags: [String: String]
        var metadata: [String: Any]
    }
    
    private struct ErrorInfo {
        let errorType: String
        let message: String
        let file: String?
        let line: Int?
        let functionName: String?
    }
    
    private struct ErrorData {
        let errorType: String
        let message: String
        let stackTrace: String
        let severity: ErrorSeverity
        let file: String?
        let line: Int?
        let functionName: String?
        let metadata: [String: Any]
        
        func toDictionary() -> [String: Any] {
            var dict: [String: Any] = [
                "errorType": errorType,
                "message": message,
                "stackTrace": stackTrace,
                "severity": severity.rawValue
            ]
            
            if let file = file {
                dict["file"] = file
            }
            
            if let line = line {
                dict["line"] = line
            }
            
            if let functionName = functionName {
                dict["functionName"] = functionName
            }
            
            if !metadata.isEmpty {
                dict["metadata"] = metadata
            }
            
            return dict
        }
    }
    
    public enum ErrorSeverity: String {
        case critical = "critical"
        case error = "error"
        case warning = "warning"
        case info = "info"
    }
}

File: sdks/android/src/main/java/com/errortracker/sdk/ErrorTracker.kt
package com.errortracker.sdk

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import com.google.gson.Gson
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Error Tracker SDK for Android
 * 
 * Usage:
 * ```
 * ErrorTracker.init(context, apiKey = "err_abc123...", apiUrl = "https://api.example.com/api")
 * ```
 */
object ErrorTracker {
    private var config: Config? = null
    private var defaultUncaughtExceptionHandler: Thread.UncaughtExceptionHandler? = null
    private val gson = Gson()
    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .writeTimeout(5, TimeUnit.SECONDS)
        .readTimeout(5, TimeUnit.SECONDS)
        .build()

    /**
     * Initialize the SDK
     */
    fun init(
        context: Context,
        apiKey: String,
        apiUrl: String = "http://localhost:3000/api",
        environment: String? = null,
        release: String? = null,
        enabled: Boolean = true,
        user: Map<String, Any>? = null,
        tags: Map<String, String>? = null,
        metadata: Map<String, Any>? = null,
    ) {
        if (config != null) {
            android.util.Log.w("ErrorTracker", "SDK already initialized")
            return
        }

        if (apiKey.isEmpty()) {
            throw IllegalArgumentException("ErrorTracker: apiKey is required")
        }

        config = Config(
            context = context,
            apiKey = apiKey,
            apiUrl = apiUrl.trimEnd('/'),
            environment = environment,
            release = release ?: getAppVersion(context),
            enabled = enabled,
            user = user ?: emptyMap(),
            tags = tags ?: emptyMap(),
            metadata = metadata ?: emptyMap(),
        )

        if (enabled) {
            setupAutomaticCapture()
        }
    }

    /**
     * Capture an error manually
     */
    fun captureError(
        error: Throwable,
        severity: ErrorSeverity = ErrorSeverity.ERROR,
        metadata: Map<String, Any>? = null,
    ) {
        val config = this.config ?: run {
            android.util.Log.w("ErrorTracker", "SDK not initialized. Call init() first.")
            return
        }

        if (!config.enabled) {
            return
        }

        val errorData = buildErrorData(error, severity, metadata, config)
        sendError(errorData, config)
    }

    /**
     * Capture a message (non-error)
     */
    fun captureMessage(
        message: String,
        severity: ErrorSeverity = ErrorSeverity.INFO,
        metadata: Map<String, Any>? = null,
    ) {
        val config = this.config ?: run {
            android.util.Log.w("ErrorTracker", "SDK not initialized. Call init() first.")
            return
        }

        if (!config.enabled) {
            return
        }

        val errorData = ErrorData(
            errorType = "Message",
            message = message,
            stackTrace = Thread.currentThread().stackTrace.joinToString("\n") { it.toString() },
            severity = severity,
            metadata = buildMetadata(config, metadata),
        )

        sendError(errorData, config)
    }

    /**
     * Set user context
     */
    fun setUser(user: Map<String, Any>) {
        config?.let {
            it.user = user
        } ?: android.util.Log.w("ErrorTracker", "SDK not initialized. Call init() first.")
    }

    /**
     * Set tags
     */
    fun setTags(tags: Map<String, String>) {
        config?.let {
            it.tags = it.tags + tags
        } ?: android.util.Log.w("ErrorTracker", "SDK not initialized. Call init() first.")
    }

    /**
     * Set metadata
     */
    fun setMetadata(metadata: Map<String, Any>) {
        config?.let {
            it.metadata = it.metadata + metadata
        } ?: android.util.Log.w("ErrorTracker", "SDK not initialized. Call init() first.")
    }

    /**
     * Enable/disable the SDK
     */
    fun setEnabled(enabled: Boolean) {
        config?.let {
            it.enabled = enabled
        } ?: android.util.Log.w("ErrorTracker", "SDK not initialized. Call init() first.")
    }

    private fun setupAutomaticCapture() {
        defaultUncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()

        Thread.setDefaultUncaughtExceptionHandler { thread, exception ->
            // Capture the error
            config?.let {
                captureError(exception, ErrorSeverity.CRITICAL)
            }

            // Call original handler
            defaultUncaughtExceptionHandler?.uncaughtException(thread, exception)
        }
    }

    private fun buildErrorData(
        error: Throwable,
        severity: ErrorSeverity,
        additionalMetadata: Map<String, Any>?,
        config: Config,
    ): ErrorData {
        val stackTrace = error.stackTraceToString()
        val errorInfo = extractErrorInfo(error, stackTrace)

        return ErrorData(
            errorType = errorInfo.errorType,
            message = errorInfo.message,
            stackTrace = stackTrace,
            severity = severity,
            file = errorInfo.file,
            line = errorInfo.line,
            functionName = errorInfo.functionName,
            metadata = buildMetadata(config, additionalMetadata),
        )
    }

    private fun extractErrorInfo(error: Throwable, stackTrace: String): ErrorInfo {
        val errorType = error.javaClass.simpleName
        val message = error.message ?: "Unknown error"

        // Try to extract file, line, and function from stack trace
        val stackLines = stackTrace.split("\n")
        var file: String? = null
        var line: Int? = null
        var functionName: String? = null

        if (stackLines.isNotEmpty()) {
            // Pattern: at com.package.Class.method(File.kt:123)
            val match = Regex("at\\s+[\\w.]+\\.[\\w]+\\(([^:]+):(\\d+)\\)").find(stackLines[0])
            if (match != null) {
                file = match.groupValues[1].split("/").last()
                line = match.groupValues[2].toIntOrNull()
                functionName = stackLines[0].substringBefore("(").substringAfterLast(".")
            }
        }

        return ErrorInfo(errorType, message, file, line, functionName)
    }

    private fun buildMetadata(
        config: Config,
        additionalMetadata: Map<String, Any>?,
    ): Map<String, Any> {
        val metadata = mutableMapOf<String, Any>(
            "os" to "Android",
            "osVersion" to Build.VERSION.RELEASE,
            "sdkVersion" to Build.VERSION.SDK_INT,
            "deviceModel" to Build.MODEL,
            "deviceManufacturer" to Build.MANUFACTURER,
            "framework" to "android",
        )

        config.environment?.let { metadata["environment"] = it }
        config.release?.let { metadata["release"] = it }

        config.user["id"]?.let { metadata["userId"] = it }
        config.user["username"]?.let { metadata["userName"] = it }

        if (config.tags.isNotEmpty()) {
            metadata["tags"] = config.tags
        }

        metadata.putAll(config.metadata)
        additionalMetadata?.let { metadata.putAll(it) }

        return metadata
    }

    private fun getAppVersion(context: Context): String? {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            packageInfo.versionName
        } catch (e: PackageManager.NameNotFoundException) {
            null
        }
    }

    private fun sendError(errorData: ErrorData, config: Config) {
        val endpoint = "${config.apiUrl}/errors/report"
        val json = gson.toJson(errorData)

        val request = Request.Builder()
            .url(endpoint)
            .post(json.toRequestBody("application/json".toMediaType()))
            .addHeader("Authorization", "Bearer ${config.apiKey}")
            .addHeader("Content-Type", "application/json")
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                android.util.Log.w("ErrorTracker", "Failed to send error: ${e.message}")
            }

            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    android.util.Log.w("ErrorTracker", "Failed to send error: ${response.code}")
                }
                response.close()
            }
        })
    }

    private data class Config(
        val context: Context,
        val apiKey: String,
        val apiUrl: String,
        val environment: String?,
        val release: String?,
        var enabled: Boolean,
        var user: Map<String, Any>,
        var tags: Map<String, String>,
        var metadata: Map<String, Any>,
    )

    private data class ErrorInfo(
        val errorType: String,
        val message: String,
        val file: String?,
        val line: Int?,
        val functionName: String?,
    )

    private data class ErrorData(
        val errorType: String,
        val message: String,
        val stackTrace: String,
        val severity: ErrorSeverity,
        val file: String? = null,
        val line: Int? = null,
        val functionName: String? = null,
        val metadata: Map<String, Any>,
    )

    enum class ErrorSeverity {
        CRITICAL, ERROR, WARNING, INFO
    }
}

